### PR TITLE
continuing to isolate v2 hashing code

### PR DESCRIPTION
**Pull request by aryairani**
- [x] let's fix test failures in `topic/rehash-codebase` before merging this; apart from that, it's okay to review now
- [ ] `gitsync22.sc.gist` is failing due to unknown explicitly referenced hash `frj8ob9ugr`
- [ ] `gitsync22.sc.history` is failing due to unknown explicitly referenced hash `0u7no051k7`

This PR:
1. moves Causal data structures from `Unison.Causal` to `Unison.Causal.Type` to avoid cyclic dependencies with hashing module, leaving just convenience functions and re-exports in the original file.
2. deletes Causal cachedRead / deserialization stuff; (this could potentially be a separate PR, but doesn't seem controversial)
3. delegates all Causal hashing to `Hashing.hashCausal`. I created new read-only patterns `One` / `Cons` / `Merge`, and rename the constructors to `UnsafeOne`, `UnsafeCons`, `UnsafeMerge`; these are used now in very limited circumstances.
4. moves most of `Hashable.hs` into `BuildHashable.hs`; Hashable is less composable now by design.
5. leaves `Unison.Hashable` in place for some bit sprinkling that Parser.hs is still doing, but removed the Hashable1 typeclass from that module. Some arbitrary external hashing library could be used here, but I didn't bother to find and integrate one.

### CODE REVIEW

File: parser-typechecker/src/Unison/Hashing/V2/Convert.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Hashing/V2/Convert.hs]: Are we planning to change this at all as part of this migration?

File: parser-typechecker/src/Unison/Hashing/V2/Hashable.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Hashing/V2/Hashable.hs]: I know you're just moving this, but does this actually need the Functor constraint? Seems like it doesn't depend on it, and the implementer of the instance should probably be able to decide their own constraints right?

I don't mind if it stays, mostly just asking out of curiousity.

File: parser-typechecker/src/Unison/Hashing/V2/Branch.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Hashing/V2/Branch.hs:25]: Maybe `hashRawBranch` to avoid confusion with the other `hashBranch` and `hashBranch0`?

**Review by ChrisPenner — COMMENTED** [review]
Looks fine to me; I'll admit I don't love having overlapping names, even in different modules, generally I find it just makes things a bit tougher to grok at a glance, but that'd be my only criticism 😄

File: parser-typechecker/src/Unison/Hashing/V2/Hashable.hs
Comment (aryairani) [reply parser-typechecker/src/Unison/Hashing/V2/Hashable.hs]: Seems like an oversight and that it should go.

File: parser-typechecker/src/Unison/Hashing/V2/Convert.hs
Comment (aryairani) [reply parser-typechecker/src/Unison/Hashing/V2/Convert.hs]: Oops, yes, this PR would change this, but I forgot to change the comment.

File: parser-typechecker/src/Unison/Hashing/V2/Branch.hs
Comment (aryairani) [reply parser-typechecker/src/Unison/Hashing/V2/Branch.hs:25]: Hm I think it's okay, this is part of a pretty different interface, and I wouldn't anticipate ever importing both packages.

But also, the fact that the type defined in this file is called `Raw` is something of a historical artifact, and could just as easily be named `Branch` as well. I think the function here is named more after the module and the concept than the data type.

File: stack.yaml
Comment (aryairani) [inline stack.yaml]: Whoops, didn't mean to check in this change.

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs
Comment (aryairani) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs]: Not sure why this shows up here, I thought someone else deleted it.

**Review by mitchellwrosen — APPROVED** [review]

File: unison-core/src/Unison/Hashable.hs
Comment (aryairani) [inline unison-core/src/Unison/Hashable.hs:42]: This is no longer used for our hashing, so we don't need to salt anything.  Perhaps we'll refactor `Parser.hs` at some point to no longer require this module.

File: parser-typechecker/src/Unison/Hashing/V2/Hashable.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Hashing/V2/Hashable.hs:18]:
```suggestion

instance BuildHashable.Hashable a => Hashable (Set a) where
```

File: parser-typechecker/src/Unison/Codebase/Causal.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Causal.hs:151]: The name `fromList` isn't really suggesting how this should be used to me.
It's also a bit weird to have a `fromList` that takes an extra argument than just the list.

```suggestion
consOntoTails :: (Applicative m, Hashable e) => e -> [Causal m h e] -> Causal m h e
```

File: parser-typechecker/src/Unison/Codebase/Causal.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Causal.hs]: What should we do here if `h` is equal to `currentHash t`?
We should probably de-dupe this right?

File: parser-typechecker/src/Unison/Codebase/Causal/Type.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Causal/Type.hs:25]: Love the descriptions, but maybe these should just be haddocks (at the very least a module doc, so it shows up in haddocks?)
I'm worried about this description, or the contained types, going out of date.
Case-in-point, the `Causal` type in the descriptions switches between `Causal a` and `Causal m a`, when in fact the type is `Causal m h a` 🤔 

---

Edit: after reading through this module I see that none of these operations exist as combinators here as the docs seem to imply, are you referring to the Patterns? Where is `sequence` defined?

Now I'm very confused 😓 ; not sure what this block of docs is trying to say.

---

Edit 2:
I didn't realize this was simply moved (in the future, notes within the PR about what changed, what moved, etc. really help with this 😄 )
I think the comments still stand, but up to you whether we make those changes now or not.

File: parser-typechecker/src/Unison/Codebase/Causal/Type.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Causal/Type.hs:34]: Not sure what this is implying; The fact that this is a type signature implies the existence of a `CommutativeSemigroup` class somewhere which I can't find in this PR. I suspect you mean it more conceptually, e.g. "You need to tell me how to merge two or more branches in order to merge", but I think I'd rather have that written in English rather than referring to classes we don't actually use.

I don't see which definition this is referring to in this module though, nothing seems to accept anything resembling a CommutativeSemigroup.

File: parser-typechecker/src/Unison/Codebase/Causal/Type.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Causal/Type.hs]: This lens is unsafe; editing the head invalidates the hash, but doesn't update it.

We can either downgrade this to a 'getter', or replace with the appropriate lens that updates the hash after updating.

File: parser-typechecker/src/Unison/Codebase/Causal/Type.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Causal/Type.hs]:
```suggestion
predecessors :: Causal m h e -> Seq (m (Causal m h e))
```

Not sure whether now is the right time to change it or not (I'll leave that up to you), but I believe we agreed `children` isn't the best name for this concept.

File: parser-typechecker/src/Unison/Codebase/Causal.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Causal.hs:86]: Is it necessary to split this module up right now? Seems like it's likely to cause us merge-conflict issues, would it be possible to do this refactor on trunk and merge it in instead?
Refactoring in a long-lived feature branch can cause a lot of headaches sometimes 😅

File: parser-typechecker/src/Unison/Hashing/V2/BuildHashable.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Hashing/V2/BuildHashable.hs:50]: I notice we have a lot of overlap between our Hashables; e.g. AFAICT all of these distinct types would have the same hash:

```
[1, 2]
(1, 2)
Set.fromList [1, 2]
```
etc.;

Should we be tagging these somehow? Or maybe prepending a `Text "Set"` to the front or something to disambiguate?

Maybe it's fine with the current implementation of things, but seems like it might be a really subtle bug just waiting to happen 😅

File: parser-typechecker/src/Unison/Hashing/V2/BuildHashable.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Hashing/V2/BuildHashable.hs:67]: Seems kinda weird to have this class in the middle of a bunch of Hashable instances; also do we not actually define any Hashable1 instances?
I'd expect to see instances for basic container types here if we actually used it.

File: parser-typechecker/src/Unison/Hashing/V2/BuildHashable.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Hashing/V2/BuildHashable.hs:1]: IIRC we wanted to only export certain things for safety, not sure quite where this implementation ended up on that,
was there any plan to restrict exports here?

**Review by ChrisPenner — COMMENTED** [review]
I left a bunch of comments, but I'll admit that I'm confused about what I'm reviewing here, and I think I ended up reviewing a bunch of code that was either just moved or was copy-pasted, a tour of this PR would be helpful for me 😄

File: parser-typechecker/src/Unison/Codebase/Causal.hs
Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Causal.hs]: No, we do that in `consDistinct`, but not in `cons`.

File: parser-typechecker/src/Unison/Codebase/Causal.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/Causal.hs:150]: Update docs to mention that the list is de-duped.

File: parser-typechecker/src/Unison/Codebase/Causal/Type.hs
Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Causal/Type.hs]: We need to merge in the change from Causal.hs and/or do this split off of trunk first and see what happens.

File: parser-typechecker/src/Unison/Codebase/Causal/Type.hs
Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Causal/Type.hs]: We think this is also already changed on trunk.

File: parser-typechecker/src/Unison/Hashing/V2/BuildHashable.hs
Comment (aryairani) [reply parser-typechecker/src/Unison/Hashing/V2/BuildHashable.hs:50]: We talked on the phone and decided the tags belong on the instances for sum types, and not necessarily to distinguish between multiple instances.

File: parser-typechecker/src/Unison/Hashing/V2/BuildHashable.hs
Comment (aryairani) [reply parser-typechecker/src/Unison/Hashing/V2/BuildHashable.hs:67]: We define them on the other side — ABT / Term / Type / DataDeclaration, etc.

File: parser-typechecker/src/Unison/Hashing/V2/Patch.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Hashing/V2/Patch.hs]: Monomorphize references to Accumulate

**Review by mitchellwrosen — APPROVED** [review]

**Review by ChrisPenner — APPROVED** [review]

**Comment by ChrisPenner** [comment]
We decided to try to get tests passing on topic/rehash before merging this. Mitchell and I will try to accomplish that while Arya is on holidays.

**Comment by mitchellwrosen** [comment]
~@ChrisPenner and I are getting ready to merge this, but one mystery remains. Actually, it remains a mystery whether it's a mystery.~

~On this branch, I discovered after a lot of printf debugging that we are actually changing how we hash branches (`Branch0`), relative to what's in `topic/rehash-codebase` right now.~

~Is that expected?~

**Comment by mitchellwrosen** [comment]
I did a little bit more debugging, and it turns out the actual problem is we are hashing a `One` causal differently.

Previously, we just used the branch hash directly. Now, we pass it and the empty list of predecessors along to the underlying hashable instance:

```haskell
instance H.Hashable Causal where
  tokens c = H.tokens $ branchHash c : Set.toList (parents c)
```

It seems fine to tumble in the hash of the empty list to a `One` causal, I just wanted to call that out. It's also easy to get the old behavior:

```haskell
instance H.Hashable Causal where
  tokens c =
    case Set.toList (parents c) of
      [] -> H.tokens [branchHash c]
      cs -> H.tokens (branchHash c : cs)
```

**Pull request merged**
